be/src/storage/lake/tablet_manager.cpp

```diff
@@ -395,24 +398,82 @@ Status TabletManager::publish_version(int64_t tablet_id, int64_t base_version, i
     return publish(&tablet, base_version, new_version, txns, txns_size);
 }
 
+static Status apply_write_log(const TxnLogPB_OpWrite& op_write, TabletMetadata* metadata) {
+    if (op_write.has_rowset() && op_write.rowset().num_rows() > 0) {
+        auto rowset = metadata->add_rowsets();
+        rowset->CopyFrom(op_write.rowset());
+        rowset->set_id(metadata->next_rowset_id());
+        metadata->set_next_rowset_id(metadata->next_rowset_id() + rowset->segments_size());
+    }
+    return Status::OK();
+}
+
+static Status apply_compaction_log(const TxnLogPB_OpCompaction& op_compaction, TabletMetadata* metadata) {
+    // It's ok to have a compaction log without input rowset and output rowset.
+    if (op_compaction.input_rowsets().empty()) {
+        DCHECK(!op_compaction.has_output_rowset() || op_compaction.output_rowset().num_rows() == 0);
+        return Status::OK();
+    }
+    if (UNLIKELY(!op_compaction.has_output_rowset())) {
+        return Status::InternalError("compaction log contains input rowset but no output rowset");
+    }
+    if (UNLIKELY(op_compaction.output_rowset().num_rows() == 0)) {
```

Comment:
Num rows maybe 0 if delete is supported.

Reply:
fixed